### PR TITLE
DirectAuth updates for Storage

### LINKED ISSUE #1207: Update OAuth and add email login for Storage
- ~~Integrate new API client after the API updates are made~~
- Migrate OAuth flows to use the Torus DirectAuth SDK instead of getting the OAuth redirect from the API.
- Allow login with MM, email, Github/Google

See the detailed login flows here: https://drive.google.com/file/d/1h09MiNOdiDopqkxnrDnWVAF1nukpU8cv/view?usp=sharing

### PR DESCRIPTION
**Pull request by RyRy79261**
closes #1207

### CODE REVIEW

**Review by FSM1 — APPROVED** [review]

File: packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  const [page, setPage] = useState<"confirmEmail" | "confirmVerificationCode">("confirmEmail")
+  const [email, setEmail] = useState<string | undefined>()
+  const { storageApiClient, login } = useStorageApi()
+  const [hasEmailResent, setHasEmailResent] = useState(false)
+  const [error, setError] = useState<string | undefined>()
+
+  const emailValidation = useMemo(() => yup.object().shape({
+    email: yup
+      .string()
+      .email("Please enter a valid email")
```
Comment (Tbaut) [inline packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx]:
```suggestion
      .email(t`Please enter a valid email`)
```

File: packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+    storageApiClient.getIdentityEmailToken({ email: values.email })
+      .then(() => {
+        setEmail(values.email)
+        setPage("confirmVerificationCode")
+        setIsSubmitEmailLoading(false)
+      }).catch ((e) => {
+        setError(t`Something went wrong!`)
+        setIsSubmitEmailLoading(false)
+        console.error(e)
+      })
```
Comment (Tbaut) [inline packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx]: nit
```suggestion
      .then(() => {
        setEmail(values.email)
        setPage("confirmVerificationCode")
      }).catch ((e) => {
        setError(t`Something went wrong! Please try again.`)
        console.error(e)
      }).finally (() => setIsSubmitEmailLoading(false))
```

File: packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+      setIsSubmitNonceLoading(false)
+    }).catch ((e) => {
+      if (e && e[0] && e[0].type === "nonce") {
+        setError(t`Verification code not correct!`)
+      } else {
+        setError(t`Something went wrong!`)
+      }
+      setIsSubmitNonceLoading(false)
+      console.error(e)
+    })
```
Comment (Tbaut) [inline packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx]: using finally here too.
```suggestion
    }).then(async (data) => {
      await login("email", { token: data || "", email })
    }).catch ((e) => {
      if (e && e[0] && e[0].type === "nonce") {
        setError(t`Verification code not correct!`)
      } else {
        setError(t`Something went wrong!`)
      }
      console.error(e)
    }).finally(() => setIsSubmitNonceLoading(false))
```

File: packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  const onResendEmail = useCallback(() => {
+    if (!email) return
+    setIsSubmitResendEmailLoading(true)
+    storageApiClient.getIdentityEmailToken({ email }).then(() => {
+      setHasEmailResent(true)
+      setIsSubmitResendEmailLoading(false)
+    }).catch ((e) => {
+      setIsSubmitResendEmailLoading(false)
+      console.error(e)
+    })
```
Comment (Tbaut) [inline packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx]: nit
```suggestion
    storageApiClient.getIdentityEmailToken({ email })
      .then(() => setHasEmailResent(true))
      .catch(console.error)
      .finally(() => setIsSubmitResendEmailLoading(false))
```

File: packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            <FormikTextInput
+              className={classes.input}
+              name="email"
+              label={t`Enter your email:`}
+              labelClassName={classes.inputLabel}
+            />
+            {error && (
+              <Typography component="p"
+                variant="body1"
+                className={classes.errorText}>{error}</Typography>
```
Comment (Tbaut) [inline packages/storage-ui/src/Components/Modules/LoginModule/PasswordlessEmail.tsx]: nit
```suggestion
              <Typography
                component="p"
                variant="body1"
                className={classes.errorText}
              >
                {error}
              </Typography>
```

File: packages/storage-ui/src/Contexts/StorageApiContext.tsx [hunk 6]
```diff
@@ -1,38 +1,83 @@
...
+    throw new Error(`${loginType} is unsupported`)
+  }
+  }
+}
 
+export type StorageUserInfo = {
+  typeOfLogin: IdentityProvider
+  email?: string
+  publicAddress?: string
+}
```
Comment (Tbaut) [inline packages/storage-ui/src/Contexts/StorageApiContext.tsx]: nit for consistancy
```suggestion
export interface StorageUserInfo {
  typeOfLogin: IdentityProvider
  email?: string
  publicAddress?: string
}
```

File: packages/storage-ui/src/Contexts/StorageApiContext.tsx [hunk 7]
```diff
@@ -215,38 +291,104 @@ const StorageApiProvider = ({ apiUrl, withLocalStorage = true, children }: Stora
...
+      })
+      setStatus("awaiting confirmation")
+      const oauthIdToken = await loginHandler.handleLoginWindow({})
+      setStatus("logging in")
+      const userInfo = await loginHandler.getUserInfo(oauthIdToken)
+
+      return { identityToken: {
+        expires: (oauthIdToken as ExtendedTorusResponse).expires_in,
+        token:  oauthIdToken.idToken || oauthIdToken.accessToken
+      }, userInfo: userInfo }
```
Comment (Tbaut) [inline packages/storage-ui/src/Contexts/StorageApiContext.tsx]: nit
```suggestion
      return {
        identityToken: {
          expires: (oauthIdToken as ExtendedTorusResponse).expires_in,
          token:  oauthIdToken.idToken || oauthIdToken.accessToken
        },
        userInfo
      }
```

File: packages/storage-ui/src/Contexts/StorageApiContext.tsx [hunk 8]
```diff
@@ -215,38 +291,104 @@ const StorageApiProvider = ({ apiUrl, withLocalStorage = true, children }: Stora
...
-  const web3Login = async () => {
-    if (!provider) return Promise.reject("No wallet is selected")
-
-    if (!isReady) {
-      const connected = await checkIsReady()
-      if (!connected) return Promise.reject("You need to allow the connection")
+  const getIdentityToken = async (
+    loginType: IdentityProvider,
+    tokenInfo?: {token: IdentityToken; email: string}
+  ): Promise<{identityToken: IdentityToken; userInfo: any}> => {
```
Comment (Tbaut) [inline packages/storage-ui/src/Contexts/StorageApiContext.tsx:297]: Any way we can type this strongly?

**Review by Tbaut — APPROVED** [review]
Working super well. Found a couple nits but that's it :)

File: packages/storage-ui/src/Contexts/StorageApiContext.tsx (hunk 8 above)
Comment (FSM1) [reply packages/storage-ui/src/Contexts/StorageApiContext.tsx:297]: I played around with it and have not managed to come up with a type definition that works. Given the tiny surface area of this type, I would say we should leave it as is.

**Pull request merged**